src/Microsoft.Data.SqlClient/tests/FunctionalTests/SslOverTdsStreamTest.cs

```diff
@@ -0,0 +1,345 @@
+        {
+            using (LimitedMemoryStream stream = new LimitedMemoryStream(buffer, maxPacketReadLength))
+            using (Stream tdsStream = CreateSslOverTdsStream(stream))
+            {
+                int offset = 0;
+                byte[] bytes = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+                for (int index = 0; index < encapsulatedPacketCount; index++)
+                {
+                    Array.Clear(bytes, 0, bytes.Length);
+                    int packetBytes = ReadPacket(tdsStream, action, bytes);
+                    Array.Copy(bytes, 0, output, offset, packetBytes);
+                    offset += packetBytes;
+                }
+                InvokeFinishHandshake(tdsStream);
+                for (int index = 0; index < passthroughPacketCount; index++)
+                {
+                    Array.Clear(bytes, 0, bytes.Length);
+                    int packetBytes = ReadPacket(tdsStream, action, bytes);
+                    Array.Copy(bytes, 0, output, offset, packetBytes);
+                    offset += packetBytes;
+                }
+            }
+
+        }
+
+        private static int ReadPacket(Stream tdsStream, Func<Stream, byte[], int, int, int> action, byte[] output)
+        {
+            int readCount;
+            int offset = 0;
+            byte[] bytes = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+            do
+            {
+                readCount = action(tdsStream, bytes, offset, bytes.Length - offset);
+                if (readCount > 0)
+                {
+                    offset += readCount;
+                }
+            }
+            while (readCount > 0 && offset < bytes.Length);
+            Array.Copy(bytes, 0, output, 0, offset);
+            return offset;
+        }
+
+        private static async Task<int> ReadPacket(Stream tdsStream, Func<Stream, byte[], int, int, Task<int>> action, byte[] output)
+        {
+            int readCount;
+            int offset = 0;
+            byte[] bytes = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+            do
+            {
+                readCount = await action(tdsStream, bytes, offset, bytes.Length - offset);
+                if (readCount > 0)
+                {
+                    offset += readCount;
+                }
+            }
+            while (readCount > 0 && offset < bytes.Length);
+            Array.Copy(bytes, 0, output, 0, offset);
+            return offset;
+        }
+
+        private static byte[] WritePackets(int encapsulatedPacketCount, int passthroughPacketCount, Action<Stream, int> action)
+        {
+            byte[] buffer = null;
+            using (LimitedMemoryStream stream = new LimitedMemoryStream())
+            {
+                //using (SslOverTdsStream tdsStream = new SslOverTdsStream(stream))
```

Comment:
Possibly, but leaving it in tells you what the following line is equivalent to if it were possible directly instead of needing reflection. I can take it out if you prefer but I prefer verbosity in tests just to make working out why they might have failed easier.

Reply:
Greate, I agree with verbosity in tests too. In my opinion, it is maybe a better choice to describe the reason for creating `CreateSslOverTdsStream` instead of leaving this code as a comment here. In addition, your function name also illustrates its duty.

Reply:
Comment removed.